File: core/packages/planner/src/trade.ts
import { commoditiesBySymbol } from "@elite-dangerous-sdk/data";

export interface TradeStation {
  name: string;
  system: string;
  distanceFromStar?: number;
}

export interface TradeCommodity {
  symbol: string;
  name: string;
  category: string;
  buyPrice: number;
  sellPrice: number;
  stock: number;
  demand: number;
  isRare: boolean;
}

export interface TradeRoute {
  from: TradeStation;
  to: TradeStation;
  commodity: TradeCommodity;
  profitPerUnit: number;
  totalProfit: number;
  cargoCapacity: number;
  distanceLy: number;
  profitPerLy: number;
}

export interface TradeRouteFilter {
  minProfitPerUnit?: number;
  maxDistance?: number;
  minCargo?: number;
  categories?: string[];
}

/**
 * Calculate profit for a potential trade route.
 */
export function calculateTradeProfit(
  buyPrice: number,
  sellPrice: number,
  cargoCapacity: number,
): { perUnit: number; total: number } {
  const perUnit = sellPrice - buyPrice;
  return {
    perUnit,
    total: perUnit * cargoCapacity,
  };
}

/**
 * Rank trade routes by profitability.
 */
export function rankTradeRoutes(
  routes: TradeRoute[],
  topN?: number,
): TradeRoute[] {
  const sorted = [...routes].sort((a, b) => b.profitPerUnit - a.profitPerUnit);
  return topN ? sorted.slice(0, topN) : sorted;
}

/**
 * Filter trade routes by criteria.
 */
export function filterTradeRoutes(
  routes: TradeRoute[],
  filter: TradeRouteFilter,
): TradeRoute[] {
  return routes.filter((r) => {
    if (filter.minProfitPerUnit && r.profitPerUnit < filter.minProfitPerUnit)
      return false;
    if (filter.maxDistance && r.distanceLy > filter.maxDistance) return false;
    if (filter.minCargo && r.cargoCapacity < filter.minCargo) return false;
    if (filter.categories && !filter.categories.includes(r.commodity.category))
      return false;
    return true;
  });
}

/**
 * Classify a commodity's trade type.
 */
export function getCommodityType(symbol: string): string {
  const commodity = commoditiesBySymbol.get(symbol);
  if (!commodity) return symbol;

  // Categorize by category
  const category = (commodity as any).category ?? "";
  const categoryNames: Record<string, string> = {
    Metals: "Metal",
    Minerals: "Mineral",
    Foods: "Food",
    Textiles: "Textile",
    Waste: "Waste",
    Chemicals: "Chemical",
    Machinery: "Machinery",
    Technology: "Technology",
    Weapons: "Weapon",
    Slaves: "Slave",
    "Legal Drugs": "Drug",
    Narcotics: "Narcotic",
  };

  return categoryNames[category] ?? category;
}

File: python/elite_dangerous_sdk/planner/trade.py
from dataclasses import dataclass
from typing import List, Optional


@dataclass
class TradeStation:
    name: str = ""
    system: str = ""
    distance_from_star: Optional[float] = None


@dataclass
class TradeCommodity:
    symbol: str = ""
    name: str = ""
    category: str = ""
    buy_price: float = 0.0
    sell_price: float = 0.0
    stock: int = 0
    demand: int = 0
    is_rare: bool = False


@dataclass
class TradeRoute:
    source: TradeStation = None
    destination: TradeStation = None
    commodity: TradeCommodity = None
    profit_per_unit: float = 0.0
    total_profit: float = 0.0
    cargo_capacity: float = 0.0
    distance_ly: float = 0.0
    profit_per_ly: float = 0.0


def calculate_trade_profit(buy_price: float, sell_price: float, cargo_capacity: float) -> dict:
    per_unit = sell_price - buy_price
    return {"per_unit": per_unit, "total": per_unit * cargo_capacity}


def rank_trade_routes(routes: List[TradeRoute], top_n: Optional[int] = None) -> List[TradeRoute]:
    sorted_routes = sorted(routes, key=lambda r: r.profit_per_unit, reverse=True)
    if top_n is not None:
        return sorted_routes[:top_n]
    return sorted_routes


def filter_trade_routes(routes: List[TradeRoute], min_profit_per_unit: float = 0) -> List[TradeRoute]:
    return [r for r in routes if r.profit_per_unit >= min_profit_per_unit]
